src/Moq/Mock.cs

```diff
@@ -259,42 +259,77 @@ private MockException TryVerifySetups(Func<Setup, MockException> verifySetup)
 
-			if (expression.IsProperty())
+			if (!expression.IsPropertyIndexer())  // guard because `.ToPropertyInfo()` doesn't (yet) work for indexers
 			{
-				VerifyGet(mock, expression, times, failMessage);
+				var property = expression.ToPropertyInfo();
+				Guard.CanRead(property);
 			}
-			else
-			{
-				var (obj, method, args) = expression.GetCallInfo(mock);
-				ThrowIfVerifyExpressionInvolvesUnsupportedMember(expression, method);
 
-				var expectation = new InvocationShape(method, args);
-				VerifyCalls(GetTargetMock(obj, mock), expectation, expression, times, failMessage);
-			}
+			var parts = expression.Split();
+			Mock.VerifyRecursive(mock, expression, parts, times, failMessage, verifyLast: (part, targetMock) =>
+			{
+				var expectation = new InvocationShape(part.Method, part.Arguments);
+				VerifyCalls(targetMock, expectation, expression, times, failMessage);
+			});
 		}
 
-		internal static void VerifyGet(Mock mock, LambdaExpression expression, Times times, string failMessage)
+		private static void VerifyRecursive(Mock mock, LambdaExpression expression, Stack<LambdaExpressionPart> parts, Times times, string failMessage, Action<LambdaExpressionPart, Mock> verifyLast)
 		{
-			var method = expression.ToPropertyInfo().GetGetMethod(true);
-			ThrowIfVerifyExpressionInvolvesUnsupportedMember(expression, method);
+			var (expr, method, arguments) = parts.Pop();
+
+			if (mock.IsDelegateMock)
+			{
+				// The expression we have is for a call on the delegate, not our
+				// delegate interface proxy, so we need to map instead to the
+				// method on that interface.
+				_ = ProxyFactory.Instance.GetDelegateProxyInterface(mock.TargetType, out method);
+			}
+
+			ThrowIfVerifyExpressionInvolvesUnsupportedMember(expr, method);
 
-			var expectation = new InvocationShape(method, new IMatcher[0]);
-			VerifyCalls(GetTargetMock(((MemberExpression)expression.Body).Expression, mock), expectation, expression, times, failMessage);
+			if (parts.Count == 0)
+			{
+				verifyLast(new LambdaExpressionPart(expr, method, arguments), mock);
+				//                                        ^^^^^^
+				// We rebuild a new part here because `method` may have been modified.
+			}
+			else
+			{
+				Mock innerMock;
+				if (mock.GetInnerMockSetups().TryFind(method, out var setup) && setup.ReturnsInnerMock(out innerMock))
+				{
+					Mock.VerifyRecursive(innerMock, expression, parts, times, failMessage, verifyLast);
+				}
+				else if (times != Times.Never())
+				{
+					// some number of invocations were expected on a inner mock that doesn't exist.
+					throw MockException.NoMatchingCalls(
+						failMessage,
+						mock.Setups.ToArrayLive(s => AreSameMethod(s.Expression, expr)),
+						mock.MutableInvocations.ToArray(),
+						expression,
+						times, 0);
+
+					bool AreSameMethod(LambdaExpression l, LambdaExpression r) =>
+						l.Body is MethodCallExpression lc && r.Body is MethodCallExpression rc && lc.Method == rc.Method;
```

Comment:
There's some small amount of code duplication here. (The same local function exists in `VerifyCalls`.)

Reply:
We'll soon need to improve `SetupCollection` so that it can identify setups matching a call expression. Best to take care of this duplication when that happens.